src/expr.jl

```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
     end
     return kwargs
 end
+
+"""
+    splitdef(ex::Expr; throw::Bool=true) -> Union{Dict{Symbol,Any}, Nothing}
+
+Split a function definition expression into its various parts including:
+
+- `:type`: Type of the function (long-form `:function`, short-form `:(=)`, anonymous `:(->)`)
+- `:name`: Name of the function (not present for anonymous functions)
+- `:params`: Parametric types defined on constructors
+- `:args`: Positional arguments of the function
+- `:kwargs`: Keyword arguments of the function
+- `:rtype`: Return type of the function
+- `:whereparams`: Where parameters
+- `:body`: Function body (not present for empty functions)
+
+All parts are optional with the exception of `:type`. When `throw=true` and the provided
+expression is not a function definition than an exception will be raised. When `throw=false`
+then `nothing` will be returned for invalid definitions.
+"""
+function splitdef(ex::Expr; throw::Bool=true)
+    def = Dict{Symbol,Any}()
+
+    function invalid_def(section)
+        if throw
+            # Using a closure ensures that `ex` contains the original full expression
+            msg = "Function definition contains $section\n$(sprint(Meta.dump, ex))"
+            Base.throw(ArgumentError(msg))
+        else
+            nothing
+        end
+    end
+
+    # long-form `:function`, short-form `:(=)`, anonymous `:(->)`
+    if !(ex.head === :function || ex.head === :(=) || ex.head === :(->))
+        return invalid_def("invalid function type `$(repr(ex.head))`")
+    end
+
+    def[:type] = ex.head
+    anon = ex.head == :(->)
```

Comment:
One can also create an anon function using

```

julia> function()
       10^2
       end
#5 (generic function with 1 method)
```

I do not see a test for it below

Reply:
Another fun test case is

```
function(::T) where T
       10^2
end
```

Reply:
I'm now testing against long-form anonymous functions